People Management v1

oguzkocer opened the pull request:
Relevant #3901. This PR introduces the first version of people management. For now, it's a WordPress.com only feature. If you have `list_users` capability for a certain site, you should see a `People` button in the `Configuration` section of the site. You can find more information about the capabilities [here](https://codex.wordpress.org/Roles_and_Capabilities#Capabilities).

When you select a user from the list, you should go into the detail page. Initially this was implemented as two different activities but we have decided to convert them into Fragments to better support tablets.

In the detail page, you should be able to change the role of the user if you have `promote_users` capability and remove a user from a blog if you have `remove_users` capability. We have decided to go with a static string array for the selectable roles for the v1 since this feature is WordPress.com only for now. There is an issue for updating this to use the roles from network #4050 to better support self-hosted blogs since they can have any number of roles.

Because we decided to go with multiple fragments in a single activity, the network calls and callbacks are generally handled in the activity. The activity will then update the DB and should invoke refresh methods of the relevant & on screen fragments.

The work done so far has been reviewed by @hypest in smaller PRs that got merged into this branch. #3364, #3916, #3970, #3991, #4051, #4086.

Here are the error messages I've used for the network calls, it was just something temporary I've written up. It'd be great to get some feedback on them: /cc @rickybanister
> Couldn't retrieve users for the site
> Couldn't update role of the user
> Couldn't remove user

For most other messages I've used the same ones we have in Calypso, so they should be mostly fine.

The work for this feature has started a very long time ago and I've been on & off on this feature which led to design & architectural changes. I've also made some mistakes myself which led to re-writes of some sections. These shouldn't affect the final result, but I wanted to put it out there nevertheless.

The only thing left should be the analytics events for the feature. @astralbodies & @daniloercoli any tips on which events we should add here? Do we have any event names we already use on the other platforms? /cc @jleandroperez 

While this PR is in review, I am going to test this feature with different emulators one more time to make sure I haven't missed anything in a previous OS or a device.

Here are some screenshots of the feature:

![1](https://cloud.githubusercontent.com/assets/662023/15358415/fcdfe694-1d0c-11e6-8904-a9eae12d185e.png)
![2](https://cloud.githubusercontent.com/assets/662023/15358417/fce56010-1d0c-11e6-853c-c0ac7abddfe3.png)
![3](https://cloud.githubusercontent.com/assets/662023/15358416/fce4d2ee-1d0c-11e6-8efa-dfc1a61572bc.png)

Needs review: @nbradbury

daniloercoli commented:
> The only thing left should be the analytics events for the feature. @astralbodies & @daniloercoli any tips on which events we should add here? Do we have any event names we already use on the other platforms? /cc @jleandroperez

Would you mind to check what events are being tracked in Calypso?

nbradbury commented:
There should be a different icon for the People feature. Right now it's the same as the Settings icon.

![people-icon](https://cloud.githubusercontent.com/assets/3903757/15410513/2a3a4c34-1de8-11e6-88d7-058cf98f09ff.png)

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/people/PeopleManagementActivity.java:
```diff
@@ -0,0 +1,243 @@
+        }
+
+        int localBlogId = BlogUtils.getBlogLocalId(WordPress.getCurrentBlog());
+        Blog blog = WordPress.getBlog(localBlogId);
```
This seems odd since it's getting the ID of the current blog in order to lookup the current blog. And the code below doesn't deal very well with the current blog being null - most likely the activity should exit in that situation.

How about this instead?
```
Blog blog = WordPress.getCurrentBlog();
if (blog == null) {
    ToastUtils.showToast(this, R.string.blog_not_found);
    finish();
    return;
}

if (savedInstanceState == null) {
    PeopleListFragment peopleListFragment = PeopleListFragment.newInstance(blog.getLocalTableBlogId());

    getFragmentManager().beginTransaction()
            .add(R.id.fragment_container, peopleListFragment, KEY_PEOPLE_LIST_FRAGMENT)
            .commit();
}

refreshUsersList(blog.getDotComBlogId(), blog.getLocalTableBlogId());
```

nbradbury commented on WordPress/src/main/java/org/wordpress/android/datasets/PeopleTable.java:
```diff
@@ -0,0 +1,144 @@
+import java.util.List;
+
+public class PeopleTable {
+    public static final String PEOPLE_TABLE = "people";
```
`PEOPLE_TABLE` can be private.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/models/Blog.java:
```diff
@@ -515,7 +517,30 @@ public void setPlanID(long planID) {
         this.planShortName = StringUtils.notNullStr(name);
     }
+
+    public String getCapabilities() {
```
I recommend following the "never return null" rule for `getCapabilities()`, ie:
```
public String getCapabilities() {
    return StringUtils.notNullStr(capabilities);
}
```

oguzkocer replied:
I honestly didn't know about this rule, I'll try to follow it from now on. For what it's worth, I was checking for the `null` case of capabilities where we used it.

nbradbury replied:
> I honestly didn't know about this rule, I'll try to follow it from now on.

I should mention that this isn't a rule for the entire app - it's just something I've found useful, to avoid having to check the result for `null` all the time.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/models/Person.java:
```diff
@@ -0,0 +1,126 @@
+    private String avatarUrl;
+    private String role;
+
+    public Person(long personID,
```
I know this follows the way we construct a few other models in the app, but I'm really not a fan of a long list of parameters in the constructor. It's too easy to mess up the order of things.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/models/Person.java:
```diff
@@ -0,0 +1,126 @@
+        return localTableBlogId;
+    }
+
+    public String getUsername() {
```
The aforementioned "never return null" rule would be good for all the string getters here.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/accounts/helpers/FetchBlogListWPCom.java:
```diff
@@ -44,6 +44,9 @@ protected void fetchBlogList(final Callback callback) {
                     site.put("isAdmin", jsonSite.get("user_can_manage"));
                     site.put("isVisible", jsonSite.get("visible"));
 
+                    // store capabilities as a text blob
```
A minor nit, but a "blob" is typically an image or some other large object, so this comment might be confusing to future devs.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/people/PeopleAdapter.java:
```diff
@@ -0,0 +1,97 @@
+    public long getItemId(int position) {
+        Person person = getItem(position);
+        if (person == null) {
+            return 0;
```
Is `0` ever a valid person ID? If so, it might be better to return `-1` when person is null.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/people/PeopleListFragment.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    @Override
+    public void onAttach(Context context) {
```
These two `onAttach` implementations only exist in order to set the `OnPersonSelectedListener`. It might be simpler to have a `setOnPersonSelectedListener` method instead.

oguzkocer replied:
I've made the change in dc829af26cbb8f39a4a722c71ce6933574f31f1a. While I was at it, I've also changed `RoleChangeDialogFragment` to follow the same practice in f0e7a3322d2df0e330ce2e87f462cea1c4e9b0b5.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/WordPressDB.java:
```diff
@@ -748,6 +760,7 @@ public boolean deleteBlog(Context ctx, int id) {
         int rowsAffected = db.delete(BLOGS_TABLE, "id=?", new String[]{Integer.toString(id)});
         deleteQuickPressShortcutsForLocalTableBlogId(ctx, id);
         deleteAllPostsForLocalTableBlogId(id);
+        PeopleTable.deletePeopleForBlog(id);
```
I think it would be a good idea to rename this to `deletePeopleForLocalBlogId` to make it clear it's a **local** blog ID.

nbradbury commented:
@oguzkocer I'm done with the code review, but I haven't looked at the UI yet - I'll give feedback on that later (if there is any).

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/people/PeopleAdapter.java:
```diff
@@ -0,0 +1,97 @@
+    private int mAvatarSz;
+
+    public PeopleAdapter(Context context, List<Person> peopleList) {
+        mAvatarSz = context.getResources().getDimensionPixelSize(R.dimen.avatar_sz_medium);
```
This should be `people_avatar_sz` to match what the layout uses.

oguzkocer replied:
My bad! The initial implementation had these sizes but we changed the design, this got overlooked while the design was updated. Thanks for catching it!

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/people/PersonDetailFragment.java:
```diff
@@ -0,0 +1,172 @@
+
+        Person person = loadPerson();
+        if (person != null) {
+            int avatarSz = getResources().getDimensionPixelSize(R.dimen.avatar_sz_large);
```
This should match the size used in the layout, which is `people_avatar_sz`.

nbradbury commented:
I hate to say this, but I found a fairly significant problem. The request to get a site's users [here](https://github.com/wordpress-mobile/WordPress-Android/pull/4100/files#diff-8bb030929880a454f38be24d51b65e15R50) returns at most 20 users, since that's the default limit applied by the [endpoint](https://developer.wordpress.com/docs/api/1.1/get/sites/%24site/users/).

So right now sites with more than 20 users only see a partial list of people. We could solve this by using the `number` query param, but it would have to be set pretty high to cover every site - and chances are we wouldn't like the bandwidth & battery cost of downloading and parsing a huge list of users.

What this really needs is infinite scroll, so users are requested 20 at a time using the `offset` query param.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/datasets/PeopleTable.java:
```diff
@@ -0,0 +1,144 @@
+        List<Person> people = new ArrayList<>();
+        String[] args = { Integer.toString(localTableBlogId) };
+        Cursor c = getReadableDb().rawQuery("SELECT * FROM " + PEOPLE_TABLE +
+                " WHERE local_blog_id=? ORDER BY display_name, user_name", args);
```
This uses case-sensitive sorting, so "Nick Bradbury" and "nick bradbury" don't appear next to each other in the people list. I think this should be case-insensitive like the web.

oguzkocer replied:
I've tried using `COLLATE NOCASE` in order to fix this and it works for a single column, but couldn't make it work for multiple column sorting. I've also read in multiple places that it only works for 26 ASCII characters and doesn't work on localized strings in many languages. So, I decided to go with `lower(display_name), lower(user_name)` even though it has a bit more overhead to it.

nbradbury commented:
It's not at all obvious that you can click on the role in person detail to change it - I only discovered it by accident. There should be some sort of UI hint that this is editable.

Also, after I change the role, the UI doesn't update to reflect the change until after the update request has completed. It would be better if the new role showed right away.

![device-2016-05-20-161358](https://cloud.githubusercontent.com/assets/3903757/15440913/06072eb8-1ea6-11e6-8fb2-b0311c39b94c.png)

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/people/PersonDetailFragment.java:
```diff
@@ -0,0 +1,172 @@
+package org.wordpress.android.ui.people;
```
The are a number of unused imports in this class.

oguzkocer commented:
> I know this follows the way we construct a few other models in the app, but I'm really not a fan of a long list of parameters in the constructor. It's too easy to mess up the order of things.

@nbradbury I've fixed this in ac1e6bd03cf8530e6a1859cbba5ddb602a27ad49. I've opted to use `personID, blogId, localTableBlogId` as constructor parameters because I don't think it makes sense to have setters for these properties. I've used the setters for the rest of them.

oguzkocer commented:
> It's not at all obvious that you can click on the role in person detail to change it - I only discovered it by accident. There should be some sort of UI hint that this is editable.

@hypest has raised a similar concern in #4086 and @rickybanister came up with [an alternative design](https://github.com/wordpress-mobile/WordPress-Android/pull/4086#issuecomment-219475599) which might make things better. However, after exploring it a bit, the solutions we came up with (code-wise) were a bit hacky. So, we decided it's worth taking this on as a separate issue. So, for now I just added it to my checklist [here](https://github.com/wordpress-mobile/WordPress-Android/issues/3901#issuecomment-217828361). If you are OK with that, I'll open a separate issue and we can maybe discuss it there?

> Also, after I change the role, the UI doesn't update to reflect the change until after the update request has completed. It would be better if the new role showed right away.

This has been fixed in d2847cf9b64950b6ef07db7ff15b4da77dd35647.

/cc @nbradbury

nbradbury commented:
> So, for now I just added it to my checklist here. If you are OK with that, I'll open a separate issue and we can maybe discuss it there?

Yes, that's fine with me - chances are we'll want to make separate issues for all non-blocking issues, since this PR is already very large.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/people/RoleChangeDialogFragment.java:
```diff
@@ -0,0 +1,151 @@
+        return roleChangeDialogFragment;
+    }
+
+    public void setOnChangeListener(OnChangeListener listener) {
```
Hey @oguzkocer , I'm pretty sure that this will introduce rotation problems, where the listener will just end up being null because the caller did not set the listener again.

oguzkocer commented:
@nbradbury I updated the people icon in my site fragment fa521fffff3eb96c0e5d911c6136d7f62b86e8d8. We have the same icon in me fragment, so I re-used that resource. I was just wondering if it's OK to leave the name as is `me_icon_my_profile` or should I try to change it to something more general. In any case, we'll break the naming convention a bit because the me fragment icons start with `me_` and my site ones start with `my_site_`.

Is there a way to have an alias for a resource by any chance? 🎅 

Might not be a big deal at all, but I have OCD against these things..

oguzkocer commented:
Hey @hypest! I've fixed the rotation issue in the role change dialog by using `EventBus` pattern rather than listener as per @nbradbury's suggestion: 3ed87e8ca0954c62d8fd458ba1618896f7e93e53. It's the final blocker for the PR, if you can take a look, that'd be great!

I've also changed the error messages a bit per [your suggestion](https://github.com/wordpress-mobile/WordPress-Android/pull/4086#discussion_r63006322) as I didn't get any other feedback about them. If we do later on, we can always change it. I only kept the `couldn't` part since it's used heavily in other error strings instead of `could not`.

hypest commented:
LGTM! :shipit:

astralbodies commented:
WOOOOHOOOO!

oguzkocer commented:
Thanks a lot for your reviews & help @nbradbury, @hypest & @rickybanister!
